### DISCUSSION TITLE
[SDK] Are there any example or notes to use a paymaster with the Rust SDK

### DISCUSSION THREAD

**Discussion by ETeissonniere** [zkSync-Community-Hub/zksync-developers#335] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

We are using the Rust SDK to generate bindings on mobile and would be interested in crafting a transaction going through a paymaster. However, our engineers are a little confused as to how to craft such a tx with the Rust SDK.

### Expected Behavior

There should be an example or quick documentation entry to use a **custom paymaster** when crafting a transaction via the Rust SDK.

### Code Example

```
            let set_mint_result = async {
                let era_provider = zk_wallet
                    .get_era_provider()
                    .map_err(|e| format!("Failed to fetch era provider: {}", e))?;

                // set parameters for mint
                let parameters = vec![format!("{:?}", zk_wallet.l2_address()), file_url];

                // send transaction for minting
                let receipt_future = era_provider
                    .send_eip712(
                        &zk_wallet.l2_wallet,
                        contract_address,
                        "safeMint(address, string)",
                        Some(parameters),
                        None,
                    )
                    .await
                    .map_err(|e| format!("Failed to send eip712 transaction: {}", e))?;

                // get receipt
                receipt_future
                    .await
                    .map_err(|e| format!("Failed to await receipt: {}", e))?
                    .ok_or_else(|| "Transaction receipt was None".to_string())
                    .map(|receipt| {
                        log_message(
                            "DEBUG",
                            &format!("Mint transaction hash {:#?}", receipt.transaction_hash),
                        );
                        receipt
                    })
            }
            .await;
```

**Comment by dutterbutter** [comment]
Hey @ETeissonniere 

You are absolutely right we should provide an example of this in the SDK example directory. Keep in mind the rust sdk is still in development. I will see if I can whip up an example for next week and post here. 

Similarly to our js sdk you will need to make use of the paymaster params defined in this [file](https://github.com/lambdaclass/zksync-web3-rs/blob/main/src/eip712/paymaster_params.rs) and pass that along with the transaction.

**Reply by ETeissonniere** [reply]
Thanks @dutterbutter, my research shows this might be possible via a combination of the above with `Eip712TransactionRequest` and passed to `send_transaction_eip712`. Will let you know if this gets resolved before you come up with the example!

**Reply by dutterbutter** [reply, marked as answer]
Hey @ETeissonniere 

I checked with the rust sdk maintainers, and paymaster support is not fully functional yet. 

If you are up for challenge and want to contribute to add support it is open source :)

**Reply by ETeissonniere** [reply]
Happy to spend some time on this with the team. Do you have any pointers to get us started?

**Reply by dutterbutter** [reply]
@ETeissonniere that is awesome 💯  🔥 👍 

I would actually first and foremost create an issue and indicate you would like to pick it up in the [zksync-web3-rs repo](https://github.com/lambdaclass/zksync-web3-rs/issues). There the maintainers will have visibility and provide more insights!

**Comment by jablonskibrittany** [comment]
To craft a transaction using a custom paymaster with the Rust SDK, you can follow these steps:

1. Define your custom paymaster contract and its interface.
2. Use the Rust SDK to interact with the paymaster contract, similar to how you interact with other contracts.
3. Pass the address of your custom paymaster contract as the `contract_address` parameter when sending a transaction.
4. Ensure that the function signature and parameters match the method you want to call on the paymaster contract.

By integrating these steps into your transaction crafting process, you can utilize a custom paymaster with the Rust SDK effectively.

**Comment by wunrorobudaniel** [comment]
To craft a transaction through a custom paymaster using the Rust SDK, you can utilize the send_eip712_with_paymaster function provided by the EraProvider interface. This allows you to specify the paymaster address and additional parameters required for the transaction. Check the Rust SDK documentation or examples for detailed usage instructions.

**Comment by sarahzitzner** [comment]
Rust SDK allows crafting transactions with custom paymasters using the send_zk_sync_tx method. Here's a basic example:

let mut tx_builder = TransactionBuilder::new();

 Add your transaction details
tx_builder
.to(your_contract_address)
.contract_call("your_contract_method", your_method_params)
.fee_token(TokenLike::Id(TokenId(0)))  Use appropriate token ID
.fee(fee_amount) // Set your desired fee
.nonce(your_nonce); // Set your nonce

 Specify your custom paymaster address
tx_builder = tx_builder.paymaster(your_paymaster_address);

 Sign the transaction
let signed_tx = zk_wallet.sign_zk_sync_transaction(tx_builder)?;

 Send the transaction
let tx_hash = zk_wallet.provider.send_zk_sync_tx(signed_tx).await?;

 Handle transaction hash
println!("Transaction sent. Hash: {:?}", tx_hash);

**Comment by deanhobson** [comment]
To use a custom paymaster when crafting a transaction via the Rust SDK, you can use the send_eip712_with_paymaster function provided by the SDK. This function allows you to specify the paymaster address, gas limit, and gas price for the transaction. Here's an example of how to use it: let era_provider = zk_wallet
    .get_era_provider()
    .map_err(|e| format!("Failed to fetch era provider: {}", e))?;

// set parameters for mint
let parameters = vec![format!("{:?}", zk_wallet.l2_address()), file_url];

// set paymaster address
let paymaster_address = Address::from_str("0x...").unwrap();

// set gas limit and gas price
let gas_limit = 100_000;
let gas_price = 1_000_000_000;

// send transaction for minting
let receipt_future = era_provider
    .send_eip712_with_paymaster(
        &zk_wallet.l2_wallet,
        contract_address,
        "safeMint(address, string)",
        Some(parameters),
        None,
        paymaster_address,
        gas_limit,
        gas_price,
    )
    .await
    .map_err(|e| format!("Failed to send eip712 transaction: {}", e))?;

// get receipt
receipt_future
    .await
    .map_err(|e| format!("Failed to await receipt: {}", e))?
    .ok_or_else(|| "Transaction receipt was None".to_string())
    .map(|receipt| {
        log_message(
            "DEBUG",
            &format!("Mint transaction hash {:#?}", receipt.transaction_hash),
        );
        receipt
    })
